java-checks/src/main/java/org/sonar/java/checks/tests/MockitoEqSimplificationCheck.java

```diff
@@ -0,0 +1,99 @@
+import org.sonar.java.checks.helpers.MethodTreeUtils;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S6068")
+public class MockitoEqSimplificationCheck extends IssuableSubscriptionVisitor {
+  private static final String MOCKITO = "org.mockito.Mockito";
+
+  private static final MethodMatchers MOCKITO_EQ_USED_IN_ARGUMENTS = MethodMatchers.or(
+    MethodMatchers.create().ofTypes(MOCKITO).names("when")
+      .addParametersMatcher(MethodMatchers.ANY).build(),
+    MethodMatchers.create().ofTypes("org.mockito.BDDMockito").names("given")
+      .addParametersMatcher(MethodMatchers.ANY).build()
+  );
+
+  private static final MethodMatchers MOCKITO_EQ_USED_IN_CONSECUTIVE_CALL = MethodMatchers.or(
+    MethodMatchers.create().ofTypes(MOCKITO).names("verify").withAnyParameters().build(),
+    MethodMatchers.create().ofTypes("org.mockito.InOrder").names("verify").withAnyParameters().build(),
+    MethodMatchers.create().ofTypes("org.mockito.stubbing.Stubber").names("when").withAnyParameters().build()
+  );
+
+  private static final MethodMatchers MOCKITO_EQ_MATCHERS = MethodMatchers.create()
+    .ofTypes("org.mockito.Matchers")
+    .names("eq").withAnyParameters().build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree mit = (MethodInvocationTree) tree;
+
+    if (MOCKITO_EQ_USED_IN_ARGUMENTS.matches(mit)) {
+      ExpressionTree argument = mit.arguments().get(0);
+      if (argument.is(Tree.Kind.METHOD_INVOCATION)) {
+        reportUselessEqUsage(((MethodInvocationTree) argument).arguments());
+      }
+    } else if (MOCKITO_EQ_USED_IN_CONSECUTIVE_CALL.matches(mit)) {
+      MethodTreeUtils.consecutiveMethodInvocation(mit).ifPresent(m -> reportUselessEqUsage(m.arguments()));
+    }
+  }
+
+  private void reportUselessEqUsage(Arguments arguments) {
+    List<MethodInvocationTree> eqs = new ArrayList<>();
+
+    for (Tree t : arguments) {
+      if (t.is(Tree.Kind.METHOD_INVOCATION) && MOCKITO_EQ_MATCHERS.matches((MethodInvocationTree) t)) {
+        eqs.add((MethodInvocationTree) t);
+      } else {
+        return;
+      }
+    }
+
+    if (!eqs.isEmpty()) {
+      reportIssue(ExpressionUtils.methodName(eqs.get(0)), String.format(
+        "Remove %s useless \"eq\" invocation and directly use the value.", eqs.size() == 1 ? "this" : "these"),
+        eqs.stream()
+          .skip(1)
+          .map(eq -> new JavaFileScannerContext.Location("", ExpressionUtils.methodName(eq)))
```

Comment:
I feel this will violate @nicolas-harraudeau-sonarsource 's rule about never having text in secondaries. Can you check?

Reply:
@m-g-sonar thanks for pinging me.
If I understand correctly we want to have secondary locations on every subsequent call to `eq()`. It seems to me that the calls to `given`, `when` and `verify` will rarely be split on many lines. If it is the case developers won't expect to navigate to `eq()` calls. They will all be visible right away.

As there is no need for navigation to these secondary locations, nor to to explain them (the primary message is clear enough), I would say that we don't need a secondary location message. Secondary locations are only there to improve issues' readability.
Does it make sense?

Reply:
Make sense. And to be clear:

```
new JavaFileScannerContext.Location("", ExpressionUtils.methodName(eq))
```

is already a message-less secondary, as `methodName` returns a tree and is used for the location.